File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums = sorted(nums)
        answer = sum(nums[:3])

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            l = i + 1
            r = len(nums) - 1

            while l < r:
                temp = nums[i] + nums[l] + nums[r]

                if temp < target:
                    l += 1
                elif temp > target:
                    r -= 1
                else:
                    return temp
                
                if abs(target - temp) < abs(target - answer):
                    answer = temp

        return answer

File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int answer = 0;
        int minGap = Integer.MAX_VALUE;
        
        for (int i = 0; i < nums.length; i++){
            int n = nums[i];
            if (i >= 1 && nums[i] == nums[i-1]){
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;
            while (left < right){
                int sum = n + nums[left] + nums[right];           
                int gap = Math.abs(target - sum);
                if (minGap > gap){
                    answer = sum;
                    minGap = gap;
                }
                if ( target < sum ){               
                    while (left < right && nums[right] == nums[right-1]){
                        right--;
                    }
                    right--;               
                } else if (sum < target) {
                    while(left < right && nums[left] == nums[left+1]){
                        left++;
                    }
                    left++;
                } else{
                    return sum;
                }
            }
        }
        return answer;
    }
}